File: aoc2024/src/day12.ts
import { Grid, linesToCharGrid, toNumberGrid } from "./utils/char-grid";
import { readInput } from "./utils/file-utils";
import { CardinalDirection, Point } from "./utils/point";

const lines = readInput("day12", false);

function expandPlot(
  dir: CardinalDirection,
  plant: string,
  start: Point,
  currentPlots: Point[],
  plot: Grid<string>
) {
  const n = start.step(dir);
  if (
    plot.isValid(n) &&
    plot.getValue(n) === plant &&
    !currentPlots.find((e) => e.equals(n))
  ) {
    currentPlots.push(n);
    buildPlot(plant, n, currentPlots, plot);
  }
}

function buildPlot(
  plant: string,
  start: Point,
  currentPlots: Point[],
  plot: Grid<string>
) {
  expandPlot(CardinalDirection.N, plant, start, currentPlots, plot);
  expandPlot(CardinalDirection.S, plant, start, currentPlots, plot);
  expandPlot(CardinalDirection.E, plant, start, currentPlots, plot);
  expandPlot(CardinalDirection.W, plant, start, currentPlots, plot);
}

function calculatePerimeter(plot: Point[]): number {
  let perimeter = 0;
  for (let i = 0; i < plot.length; i++) {
    let p = 4;
    if (plot.find((e) => e.equals(plot[i].step(CardinalDirection.N)))) p--;
    if (plot.find((e) => e.equals(plot[i].step(CardinalDirection.S)))) p--;
    if (plot.find((e) => e.equals(plot[i].step(CardinalDirection.E)))) p--;
    if (plot.find((e) => e.equals(plot[i].step(CardinalDirection.W)))) p--;
    perimeter += p;
  }
  return perimeter;
}

function buildPlots(garden: Grid<string>): Point[][] {
  const visited = new Set<string>();
  const plots = new Array<Point[]>();

  for (let y = 0; y < garden.grid.length; y++) {
    for (let x = 0; x < garden.grid[y].length; x++) {
      const p = new Point(x, y);
      if (!visited.has(p.toString())) {
        const plant = garden.getValue(p)!;
        const currentPlots = new Array<Point>();
        currentPlots.push(p);
        buildPlot(plant, p, currentPlots, garden);
        currentPlots.forEach((point) => {
          visited.add(point.toString());
        });
        plots.push(currentPlots);
        console.log(`Plant ${plant} : ${currentPlots.join(" ")}`);
      }
    }
  }
  return plots;
}

function part1() {
  const garden = linesToCharGrid(lines);

  const plots = buildPlots(garden);

  let sum = 0;
  for (let i = 0; i < plots.length; i++) {
    const perimeter = calculatePerimeter(plots[i]);
    const price = perimeter * plots[i].length;
    sum += price;
    console.log(
      `perimeter is ${perimeter}, price is ${price} for ${plots[i].join(" ")}`
    );
  }
  console.log(`total price ${sum}`);
}

function countSides(plot: Point[], garden: Grid<string>): number {
  let count = 0;

  // look for top and bottom sides
  for (let y = 0; y < garden.grid.length; y++) {
    // get all plots on this row in order
    const row = plot.filter((e) => e.y === y).sort((a, b) => a.x - b.x);

    let lastTopEdge;
    let lastBottomEdge;
    for (let x = 0; x < row.length; x++) {
      const p = row[x];
      const above = p.step(CardinalDirection.N);
      const below = p.step(CardinalDirection.S);

      const isTopEdge = garden.getValue(above) !== garden.getValue(p);
      const isBottomEdge = garden.getValue(below) !== garden.getValue(p);

      if (isTopEdge) {
        if (lastTopEdge != p.x - 1) {
          // console.log(`${garden.getValue(p)} Starting new top edge with ${p}`);
          count += 1;
        }
        lastTopEdge = p.x;
      }
      if (isBottomEdge) {
        if (lastBottomEdge != p.x - 1) {
          // console.log(`${garden.getValue(p)} Starting new bottom edge with ${p}`);
          count += 1;
        }
        lastBottomEdge = p.x;
      }
    }
  }

  // look for left and right sides
  for (let x = 0; x < garden.grid[0].length; x++) {
    // get all plots on this column in order
    const col = plot.filter((e) => e.x === x).sort((a, b) => a.y - b.y);

    let lastLeftEdge;
    let lastRightEdge;
    for (let y = 0; y < col.length; y++) {
      const p = col[y];
      const left = p.step(CardinalDirection.W);
      const right = p.step(CardinalDirection.E);

      const isLeftEdge = garden.getValue(left) !== garden.getValue(p);
      const isRightEdge = garden.getValue(right) !== garden.getValue(p);

      if (isLeftEdge) {
        if (lastLeftEdge !== p.y - 1) {
          //   console.log(`${garden.getValue(p)} Starting new left edge with ${p}`);
          count += 1;
          // } else {
          //   console.log(`${garden.getValue(p)} Continuing left edge with ${p}`);
        }

        lastLeftEdge = p.y;
      }
      if (isRightEdge) {
        if (lastRightEdge !== p.y - 1) {
          //   console.log(
          //     `${garden.getValue(p)} Starting new right edge with ${p}`
          //   );
          count += 1;
          // } else {
          //   console.log(`${garden.getValue(p)} Continuing right edge with ${p}`);
        }
        lastRightEdge = p.y;
      }
    }
  }

  return count;
}

function part2() {
  const garden = linesToCharGrid(lines);

  const plots = buildPlots(garden);

  let sum = 0;
  for (let i = 0; i < plots.length; i++) {
    const sides = countSides(plots[i], garden);
    const price = sides * plots[i].length;
    sum += price;
    console.log(
      `${garden.getValue(plots[i][0])} sides is ${sides}, price is ${price} for ${plots[i].join(" ")}`
    );
  }
  console.log(`total price ${sum}`);
}

part2();

File: aoc2024/src/python/day12.py
from collections import defaultdict
import time
from typing import Set, Tuple, List
from aoc_utils import  runIt, PuzzleInput, splitInts, Grid, Point, cardinal_directions, add, North, South, East, West
import functools
import math
import re
from itertools import chain, combinations, permutations, product
import sys


testInput = r"""AAAA
BBCD
BBCC
EEEC"""

testInput2 = r"""RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE"""
input = PuzzleInput('input/day12.txt', testInput2)
lines = input.getInputLines(test=False)
grid = Grid(lines)

def build_region(p:Point, visited:Set[Point]) -> Set[Point]:
    if p in visited:
        return set()
    
    visited.add(p)
    region = {p}
    for n in [n for n in grid.neighbors(p, cardinal_directions) if grid[n] == grid[p]]:
        region.update(build_region(n, visited))

    return region

def perimeter(region:Set[Point]) -> int:
    # Count all sides where the neighbor on that side is not also in the region
    return len([add(p, d)
                for p in region
                for d in cardinal_directions
                if add(p,d) not in region])

def sub_sides(points, dir1, dir2, region:Set[Point]) -> int:
    in1 = False
    in2 = False
    count = 0
    for p in points:
        if p not in region:
            in1 = in2  = False
        else:
            if not in1 and add(p, dir1) not in region:
                # starting a new side
                in1 = True
                count += 1
            elif in1 and add(p, dir1) in region:
                # ending a side
                in1 = False

            if not in2 and add(p, dir2) not in region:
                in2 = True
                count += 1
            elif in2 and add(p, dir2) in region:
                in2 = False
    return count

def sides(region:Set[Point]) -> int:
    count = 0

    ys = list(map(lambda p: p[1], region))
    xs = list(map(lambda p: p[0], region))
    xrange = range(min(xs), max(xs) + 1)
    yrange = range(min(ys), max(ys) + 1)

    # look for top and bottom edges by sorting the region into rows
    count += sum([sub_sides([(x, y) for x in xrange], North, South, region) for y in yrange])

    # look for right and left edges by sorting the region into columns
    count += sum([sub_sides([(x, y) for y in yrange], East, West, region) for x in xrange])

    return count

def part1():
    visited = set()

    regions = [r for r in [
        build_region(p, visited) for p in grid
    ] if len(r) > 0]

    price = sum([len(r) * perimeter(r) for r in regions])

    print(f'total price {price}')


def part2():
    visited = set()

    regions = [r for r in [
        build_region(p, visited) for p in grid
    ] if len(r) > 0]

    price = sum([len(r) * sides(r) for r in regions])

    print(f'total price {price}')

runIt(part1, part2)
